chainntnfs/bitcoindnotify/bitcoind.go

```diff
@@ -0,0 +1,765 @@
+}
+
+// Stop shutsdown the BtcdNotifier.
+func (b *BitcoindNotifier) Stop() error {
+	// Already shutting down?
+	if atomic.AddInt32(&b.stopped, 1) != 1 {
+		return nil
+	}
+
+	// Shutdown the rpc client, this gracefully disconnects from btcd, and
+	// cleans up all related resources.
+	b.chainConn.Stop()
+
+	close(b.quit)
+	b.wg.Wait()
+
+	// Notify all pending clients of our shutdown by closing the related
+	// notification channels.
+	for _, spendClients := range b.spendNotifications {
+		for _, spendClient := range spendClients {
+			close(spendClient.spendChan)
+		}
+	}
+	for _, confClients := range b.confNotifications {
+		for _, confClient := range confClients {
+			close(confClient.finConf)
+			close(confClient.negativeConf)
+		}
+	}
+	for _, epochClient := range b.blockEpochClients {
+		close(epochClient.epochChan)
+	}
+
+	return nil
+}
+
+// blockNtfn packages a notification of a connected/disconnected block along
+// with its height at the time.
+type blockNtfn struct {
+	sha    *chainhash.Hash
+	height int32
+}
+
+// notificationDispatcher is the primary goroutine which handles client
+// notification registrations, as well as notification dispatches.
+func (b *BitcoindNotifier) notificationDispatcher(currentHeight int32) {
+out:
+	for {
+		select {
+		case cancelMsg := <-b.notificationCancels:
+			switch msg := cancelMsg.(type) {
+			case *spendCancel:
+				chainntnfs.Log.Infof("Cancelling spend "+
+					"notification for out_point=%v, "+
+					"spend_id=%v", msg.op, msg.spendID)
+
+				// Before we attempt to close the spendChan,
+				// ensure that the notification hasn't already
+				// yet been dispatched.
+				if outPointClients, ok := b.spendNotifications[msg.op]; ok {
+					close(outPointClients[msg.spendID].spendChan)
+					delete(b.spendNotifications[msg.op], msg.spendID)
+				}
+
+			case *epochCancel:
+				chainntnfs.Log.Infof("Cancelling epoch "+
+					"notification, epoch_id=%v", msg.epochID)
+
+				// First, close the cancel channel for this
+				// specific client, and wait for the client to
+				// exit.
+				close(b.blockEpochClients[msg.epochID].cancelChan)
+				b.blockEpochClients[msg.epochID].wg.Wait()
+
+				// Once the client has exited, we can then
+				// safely close the channel used to send epoch
+				// notifications, in order to notify any
+				// listeners that the intent has been
+				// cancelled.
+				close(b.blockEpochClients[msg.epochID].epochChan)
+				delete(b.blockEpochClients, msg.epochID)
+
+			}
+		case registerMsg := <-b.notificationRegistry:
+			switch msg := registerMsg.(type) {
+			case *spendNotification:
+				chainntnfs.Log.Infof("New spend subscription: "+
+					"utxo=%v", msg.targetOutpoint)
+				op := *msg.targetOutpoint
+
+				if _, ok := b.spendNotifications[op]; !ok {
+					b.spendNotifications[op] = make(map[uint64]*spendNotification)
+				}
+				b.spendNotifications[op][msg.spendID] = msg
+				b.chainConn.NotifySpent([]*wire.OutPoint{&op})
+			case *confirmationsNotification:
```

Comment:
Reminder that this file should be updated to use the new shared notification struct added in a recent PR.